### PR TITLE
[risk=low][RW-9199] Upgrade React to v17

### PR DESCRIPTION
**Pull request by jmthibault79**
Updated to React 17.
React 17 makes changes to [how event delegation works](https://reactjs.org/blog/2020/10/20/react-v17.html#changes-to-event-delegation).
This change to event delegation necessitated that we upgrade to at least Prime React 7, because this is the earliest version that handles event delegation in a way that is compatible with React 17.

Updating our version of PrimeReact involved updating the Growl component to the Toast Component, changing how dropdown's attach their options to the dom, and adjusting many tests to account for selector discrepancies. 

---
**PR checklist**

- [ ] This PR meets the Acceptance Criteria in the JIRA story
- [ ] The JIRA story has been moved to Dev Review
- [ ] This PR includes appropriate unit tests
- [ ] I have added explanatory comments where the logic is not obvious
- [ ] I have run and tested this change locally, and my testing process is described here
- [ ] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later
- [ ] If this includes an API change, I have run the E2E tests on this change against my local server with [yarn test-local](https://github.com/all-of-us/workbench/blob/master/e2e/README.md#examples) because this PR won't be covered by the CircleCI tests 
- [ ] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer in [Slack](https://pmi-engteam.slack.com/archives/C02MWP2RN5P)
- [ ] If this change impacts deployment safety (e.g. removing/altering APIs which are in use) I have documented these in the description
- [ ] If this includes an API change, I have updated the appropriate Swagger definitions and updated the appropriate API consumers
  * AoU UI
  * [Perf tests](https://github.com/broadinstitute/mcnulty/blob/develop/src/test/scala/services/AoU.scala)
  * [Researcher Directory export](https://github.com/all-of-us/workbench/wiki/Researcher-Directory-(RDR-export))
  * Cron tasks - for Offline*Controllers
  * SumoLogic - for EgressAlert

### CODE REVIEW

File: ui/package.json [hunk 1]
```diff
@@ -70,7 +70,7 @@
     "react-router-dom": "^5.1.2",
     "react-select": "^3.0.8",
     "react-spring": "8.0.27",
-    "react-switch": "^4.1.0",
+    "react-switch": "^6",
```
Comment (jmthibault79) [inline ui/package.json:73]: Confirmed that `Toggle` still works as expected (our only usage of react-switch)

Comment (jmthibault79) [reply ui/package.json:73]: (upgraded to 6 because version 4 and 5 list react 16 as a peer dependency)

File: ui/yarn.lock [hunk 2]
```diff
@@ -6620,42 +6582,33 @@ object-keys@^1.1.1:
...
-    es-abstract "^1.18.2"
+    define-properties "^1.1.4"
+    has-symbols "^1.0.3"
+    object-keys "^1.1.1"
 
-object.entries@^1.1.5:
-  version "1.1.5"
-  resolved "https://registry.yarnpkg.com/object.entries/-/object.entries-1.1.5.tgz#e1acdd17c4de2cd96d5a08487cfb9db84d881861"
-  integrity sha512-TyxmjUoZggd4OrrU1W66FMDG6CuqJxsFvymeyXI51+vQLN67zYfZseptRge703kKQdo4uccgAKebXFcRCzk4+g==
+object.entries@^1.1.1, object.entries@^1.1.5:
```
Comment (jmthibault79) [inline ui/yarn.lock:6595]: a few of these patch-version upgrades are not strictly necessary.  LMK if you want me to revert

**Comment by jmthibault79** [comment]
[e2e failures](https://app.circleci.com/pipelines/github/all-of-us/workbench/23194/workflows/38d8451c-ee08-403d-9496-47f0effac6d6/jobs/249019) make me suspicious that the react-switch change is the problem.  I'll try reverting.

**Review by Peter-Lavigne — DISMISSED** [review]

File: ui/package.json (hunk 1 above)
Comment (jmthibault79) [reply ui/package.json]: this apparently broke a few e2e tests, so I reverted it.  Our existing version 4.1.0 `Toggle`s continue to work in React 17, regardless of the peer dependency claim.

**Comment by jmthibault79** [comment]
our original version of react-switch (4.1.0) continues to work fine here, so hopefully this resolves the e2e failures.

**Comment by jmthibault79** [comment]
6 e2e failures... trying again but not feeling confident

**Comment by jmthibault79** [comment]
4 [e2e failures](https://app.circleci.com/pipelines/github/all-of-us/workbench/23196/workflows/37be0fba-18cc-43e9-ba18-d2741eb63270/jobs/249035): 3 of
```
TimeoutError: waiting for XPath `//*[contains(concat(" ", normalize-space(@class), " "), " p-tieredmenu ") or contains(concat(" ", normalize-space(@class), " "), " p-menu ")][contains(concat(" ", normalize-space(@class), " "), " p-menu-overlay-visible ")][.//*[@role="menuitem"]]` failed: timeout 60000ms exceeded
```

**Comment by jmthibault79** [comment]
The e2e tests have caught a product failure: the Add Criteria button no longer works with this upgrade.

From some brief debugging, I see that cohort-criteria-menu's `onClickOutside()` is getting triggered, and this closes the menu immediately after opening it.  Why?

**Comment by jmthibault79** [comment]
This might be why: https://blog.saeloun.com/2021/07/08/react-17-event-delagation.html

File: e2e/app/component/card/data-resource-card.ts [hunk 3]
```diff
@@ -84,8 +84,13 @@ export default class DataResourceCard extends BaseCard {
     }
 
     const dataTableRows = await datatable.getRowCount();
+
     if (dataTableRows >= 1) {
-      return datatable;
+      const firstRow = await datatable.getRow(1);
```
Comment (evrii) [inline e2e/app/component/card/data-resource-card.ts:89]: Our datatables used to be blank when no filters matched:
image
But now they show an empty message which is a row:
image

We can customize the text of this message, if we would like.

File: ui/src/app/components/resource-list.tsx [hunk 4]
```diff
@@ -221,14 +221,15 @@ export const ResourceList = fp.flow(withCdrVersions())((props: Props) => {
...
+      <div data-test-id='resources-table' style={{ flex: 1 }}>
         {tableData?.length > 0 && (
           <DataTable
+            filterDisplay='row'
             data-test-id='resource-list'
             value={tableData}
-            scrollable={true}
             sortMode='multiple'
             paginator
+            breakpoint='0px'
```
Comment (evrii) [inline ui/src/app/components/resource-list.tsx:232]: The new version of the datatable component has two modes: stack and scrolled. It defaults to stacked, but setting the breakpoint to 0px prevents the alternate view. I am up for having a more detailed conversation, if people are interested.

File: ui/src/app/components/spinners.tsx [hunk 5]
```diff
@@ -35,7 +35,7 @@ export const Spinner = ({ dark = false, size = 72, style = {}, ...props }) => {
       viewBox='0 0 72 72'
       width={size}
       height={size}
-      style={{ animation: '1s linear infinite spin', ...style }}
+      style={{ animation: '1s linear infinite rotation', ...style }}
```
Comment (evrii) [inline ui/src/app/components/spinners.tsx:38]: Fixes the non-spinning spinner issue.

File: e2ev2/tests/workspace-analysis.test.js [hunk 6]
```diff
@@ -24,14 +24,8 @@ browserTest('create an application', async browser => {
...
-  const applicationListDropdown = await page.waitForSelector('input[aria-label="Application List Dropdown"]')
-  /*
-  * Since Prime React uses attaches the aria-label of its Dropdown element to an
-  * invisible element, Puppeteer's click method does not work as intended.
-  * More details can be found here:
-  * https://stackoverflow.com/a/66537619
-  * */
-  await applicationListDropdown.evaluate(b => b.click())
+  const applicationListDropdown = await page.waitForSelector('#application-list-dropdown')
+  await applicationListDropdown.click()
```
Comment (jmthibault79) [inline e2ev2/tests/workspace-analysis.test.js:28]: is this change applicable elsewhere in e2ev2?

Comment (evrii) [reply e2ev2/tests/workspace-analysis.test.js]: No, I do not believe so. It seems like the user-create-test was doing something similar. I ended up switching this to the approach that I found there, and it worked. Keeping that change for consistency.

**Review by jmthibault79 — COMMENTED** [review]
Github won't let me Approve because I'm still the author of this PR :)

But everything I've seen makes sense.  Thanks for putting in all this work!

When making side-by-side comparisons I do still see some styling differences (though some are improvements) and I have some ideas about how to proceed there, but it's a longer discussion.  I would prefer not to merge just yet.

File: e2ev2/src/circle-run-tests.sh [hunk 7]
```diff
@@ -31,8 +31,7 @@ set +e
 export FAILED_TESTS_LOG=failed-tests.txt
 # This should do the right thing. If $FAILED_TESTS is empty, nothing is specified, so Jest runs
 # all tests.
-yarn test $FAILED_TESTS \
-  --reporters=jest-silent-reporter --reporters=./src/failure-reporter.js
```
Comment (dmohs) [inline e2ev2/src/circle-run-tests.sh:35]: Failure reporter is needed to ensure re-runs don't run every test again. I think you should revert this and we can isolate it into its own PR.

File: e2ev2/tests/user-create.test.js [hunk 8]
```diff
@@ -39,7 +39,10 @@ browserTest('create user', async browser => {
     .then(eh => eh.evaluate(e => e.parentNode.parentNode.click()))
   await page.waitForSelector('[role="option"][aria-label^="Project Personnel"]')
     .then(eh => eh.click())
-  await tu.jsClick(page, '[role="button"][aria-label="Next"]')
+  await page.waitForFunction(() =>
+      document.querySelector('[role="button"][aria-label="Next"]').style.cursor !== 'not-allowed')
+
+  await page.click( '[role="button"][aria-label="Next"]')
```
Comment (dmohs) [inline e2ev2/tests/user-create.test.js:45]: Consider:
```suggestion
  await Promise.resolve('[role="button"][aria-label="Next"]').then(sel => {
    page.waitForFunction(sel => document.querySelector(sel).style.cursor !== 'not-allowed', sel)
    return sel
  }).then(sel => page.click(sel))
```

File: ui/src/app/pages/login/account-creation/account-creation-institution.spec.tsx [hunk 9]
```diff
@@ -133,6 +133,8 @@ it('should reset role value & options when institution is selected', async () =>
     originalEvent: undefined,
     value: 'Broad',
     target: { name: '', id: '', value: 'Broad' },
+    stopPropagation: () => {},
+    preventDefault: () => {},
```
Comment (dmohs) [inline ui/src/app/pages/login/account-creation/account-creation-institution.spec.tsx:137]: Consider something like:
```javascript
const eventDefaults = {stopPropagation: () => undefined, preventDefault: () => undefined}
// ...
institutionDropdown.props.onChange({...eventDefaults,
  originalEvent: undefined,
  value: 'Broad',
  target: { name: '', id: '', value: 'Broad' },
})
```

**Review by dmohs — APPROVED** [review]
Solid.

**Pull request merged**